lib/dao.js

```diff
@@ -2581,6 +2584,7 @@ DataAccessObject.prototype.save = function(options, cb) {
 
           Model.notifyObserversOf('persist', context, function(err) {
```

Comment:
Shouldn't `ctx` be passed into the callback on this line ... and 2 lines below, `ctx.data` be used instead of `context.data`? ... If not, can you explain the difference between the two variables and why this scenario is different from the rest?

Reply:
@virkt25 I guess `ctx` is only added to `loaded` hook but not `persisted` hook.

Reply:
Good question!

Let's take a look at the code:

```js
const context = {
  // ...
};

Model.notifyObserversOf('persist', context, function(err, ctx) {
  if (err) return cb(err);
  data = context.data;
  // etc.
});
```

`notifyObserversOf` accepts a context argument which is eventually forwarded as the second argument of the callback.

In this particular case, `ctx === context`.

The reason why I cannot use `context` in some places below is that those places are defining another `context` variable inside the callback. Conceptually:

```ts
const context = {
  // ...
};

Model.notifyObserversOf(..., function(err, ctx) {
  // ...
  const context = {
    // ...
  };
  Model.notifyObserversOf(..., context, ...);
});
```

What a mess :(

What do you think - should I change all places modified by this pull request to always use `ctx` from the second callback argument of `notifyObserversOf` for consistency?

Reply:
I modified `notifyObserversOf` to accept `ctx` as the second argument in most places.